lib/Dialect/StableHLO/Transforms/WrapUnderManualComputation.cpp

```diff
@@ -0,0 +1,195 @@
+  targetBlock.addArguments(argumentTypes, argumentLocations);
+
+  // Migrate all the ops currently in funcOp into the manualComputationOp
+  // region. This will also copy func.ReturnOp.
+  mlir::Block::iterator it = ++mlir::Block::iterator(manualComputationOp);
+  targetBlock.getOperations().splice(
+      targetBlock.end(), sourceBlock.getOperations(), it, sourceBlock.end());
+
+  // Create a new func.ReturnOp in the original func.funcOp that takes the
+  // manualComputationOp as it's operand.
+  builder.setInsertionPointAfter(manualComputationOp);
+  builder.create<mlir::func::ReturnOp>(builder.getUnknownLoc(),
+                                       manualComputationOp->getResults());
+
+  // Update old arguments with new arguments inside of the
+  // manualComputationBlock.
+  for (uint32_t i = 0; i < sourceBlock.getArguments().size(); i++) {
+    auto oldArg = sourceBlock.getArgument(i);
+    auto newArg = targetBlock.getArgument(i);
+
+    for (auto &targetOp : targetBlock.getOperations()) {
+      for (auto &operand : targetOp.getOpOperands()) {
+        if (operand.get() == oldArg) {
+          operand.set(newArg);
+        }
+      }
+    }
+  }
+
+  // Update all func.return ops in manualComputationOp with sdy.return op.
+  // This is because the manualComputationOp requires a sdy.returnOp.
+  for (Operation &op :
+       llvm::make_early_inc_range(targetBlock.getOperations())) {
+    if (!mlir::isa<mlir::func::ReturnOp>(op)) {
+      continue;
+    }
+
+    mlir::func::ReturnOp returnOp = mlir::cast<mlir::func::ReturnOp>(op);
+    builder.setInsertionPoint(returnOp);
+    builder.create<mlir::sdy::ReturnOp>(builder.getUnknownLoc(),
+                                        returnOp->getOperands());
+    returnOp->erase();
+  }
+
+  return mlir::success();
+}
+
+class WrapUnderManualComputationPass
+    : public impl::WrapUnderManualComputationPassBase<
+          WrapUnderManualComputationPass> {
+public:
+  using impl::WrapUnderManualComputationPassBase<
+      WrapUnderManualComputationPass>::WrapUnderManualComputationPassBase;
+
+  void runOnOperation() final {
+    mlir::ModuleOp rootModule = getOperation();
+    MLIRContext *context = rootModule.getContext();
+    mlir::OpBuilder builder(context);
+    mlir::PassManager pm(context);
+
+    // Check if the graph is already solved by shardy. If so, we will remove
+    // all sdy tensor shardings from the arguments and results.
+    if (shardy_utils::isGraphSolved(rootModule)) {
+      rootModule.walk([&](func::FuncOp funcOp) {
+        shardy_utils::removeSdyTensorShardings(context, funcOp);
+      });
+
+      return;
+    }
+
+    // If the module has gspmd annotations, we skip this pass.
+    bool gspmdAnnotationsExist =
+        shardy_utils::gspmdAnnotationsExist(rootModule);
+    if (gspmdAnnotationsExist) {
+      rootModule.emitWarning("Wrapping under manual computation pass does not "
+                             "support GSPMD annotated module for now.\n");
```

Comment:
Nit: `\n' not required.
```suggestion
                             "support GSPMD annotated module for now.");
```